### ISSUE TITLE
[Extractor] Azure is changing id of API schema on changes, leads to multiple schemas deployed and errors

### ISSUE THREAD

**Issue by bremnes** [Azure/azure-api-management-devops-resource-kit#241]
This issue is rooted in the way Azure Portal handles API changes.

For Swagger/OpenAPI:
* Operations/definitions are changed via json edit (all operations -> frontend -> edit): **schema id is changed**
* Operations/definitions are changed by re-importing definition: **schema id is changed**
* Operations are changed by using the gui to edit display name, query parameters, request, response, etc: **schema id is not changed**
* Definitions are changed by using the gui: **schema id is not changed**

When the schema id is changed, Azure either renames the old one, or creates a new one and deletes the old one. (By using the rest api we can see that there is only one schema registered on that API and what the id is.)

### Extraction
The consequences of this is that the schema id might, or might not, be changed based on what kind of "mode" the API developer chose to work in. Let's say for this scenario that the id changed.

When extracting a new version with a changed schema id the api arm template it overwritten in its entirety and all reference to the old schema id is lost from the source (except versioning of course).

However, when we deploy the ARM templates the old schema still remains, so now there are two schemas registered on the API.

### Problem ensues

Now that there are two schemas registered on the same api this leads to some problems.

1. When running the extractor from that environment there are multiple schemas in the arm templates.

2. The open api specification no longer contains the definitions when editing in the api management instance. Instead it contains a property on the end saying
```
"x-ms-export-notes": ["Definitions/body parameters are not exported since this API references multiple schemas with Swagger definitions as opposed to one. Schemas referenced: 5d7b780422ea5f19a4c52e2e,5d7b76930fde42181808ebab"]
```

3. Developer portal no longer has the correct swagger definition, same "error" as in 2) where definitions are missing.

### API still works

The operations and definitions in the API are still present and the API works as expected. The only bug here is that it's not possible to see the client contract in a friendly format such as swagger.

### How to recreate

* Create a petstore API using https://petstore.swagger.io/v2/swagger.json.
* Run extractor
* Edit the API using either the web interface or by editing the json manually
* Deploy the ARM template extracted earlier
* Either call the rest api https://management.azure.com/subscriptions/[SUBSCRIPTION-ID]/resourceGroups/[RESOURCE-GROUP-NAME]/providers/Microsoft.ApiManagement/service/[API-MANAGEMENT-INSTANCE-NAME]/apis/[API-ID]/schemas?api-version=2019-01-01 or re-run the extractor to see schema changes

### Solution (?)
Are there instances where there should be multiple schemas?

Can we in the extractor give the schema a hardcoded value instead of the semi-random generated in Azure? And throw an error if there are multiple schemas?

So for instance use `<api-name>-schema`:
```
...
"name": "[concat(parameters('ApimServiceName'), '/swagger-petstore/swagger-petstore-schema')]",
"type": "Microsoft.ApiManagement/service/apis/schemas",
...
```

as opposed to the one generated currently:
```
...
"name": "[concat(parameters('ApimServiceName'), '/swagger-petstore/5d7b780422ea5f19a4c52e2e')]",
"type": "Microsoft.ApiManagement/service/apis/schemas",
...
```

**Comment by bremnes** [comment]
@lucashuet93 , @miaojiang , any thoughts on this?

Ideally it would be interesting to figure out why the APIM engineers have chosen this approach. It doesn't seem CI/CD friendly, but hopefully there is a reason behind it.

**Comment by miaojiang** [comment]
Triaged.

**Comment by RupengLiu** [comment]
@bremnes Thanks for reporting, we will have to find other approaches to fix this problem

**Comment by elsewhat** [comment]
Experiencing the same issue with "Definitions/body parameters are not exported since this API references multiple schemas with OpenAPI components as opposed to one." and am looking for a solution or workaround. 

One of the key responsibilities of an API Management portal is to provide the OpenAPI specification to clients. This allows the client to for example generate client stubs.  Now, this is not possible as the data definitions are no longer present in OpenAPI specification downloaded from APIM. Due to the impact this issue has on one of the core responsibilities of APIM, I would have like the resolution to be prioritized higher than currently is the case.

In our case, we've been running an API-first strategy where the initial version of the OpenAPI specification has been uploaded to APIM with mock data policy enabled. As we've iterated over the implementation of the individual endpoints, there will of course be adjustments and this has been performed by updating the OpenAPI specification and re-importing to APIM. 

This issue has the following effects for us:
1. The OpenAPI specification expose through our API portal based on APIM cannot be trusted as it's incomplete
2. A separate channel for distributing the OpenAPI specification must be establish (and take into account that different dev, test, qa and prod APIM environment might have slight for endpoint under development)
3. Improved complexity for client stub generation 
4. Planned tooling for exploring the API through the command line cannot use APIM as a source

**Comment by bremnes** [comment]
@elsewhat, what we did was to create a powershell script cleaning up the unused schemas so that there is only one remaining - leaving export and other things working as normal. We initially planned on using the cleanup script as a part of our deployment pipeline, but due to locks for safety reasons we ended up running it manually from time to time instead.

**Comment by elsewhat** [comment]
@bremnes thanks for the update! Is it possible to get access to the powershell script or at least identify which powershell command used for the cleaning (assume it one of the [Azure API Management Cmdlets](https://docs.microsoft.com/en-us/powershell/module/azurerm.apimanagement/?view=azurermps-6.13.0))

**Comment by bremnes** [comment]
@elsewhat, I've re-created the script using the Az cmdlets at [bremnes/azure-api-management-redundant-schemas](https://github.com/bremnes/azure-api-management-redundant-schemas). As this is a workaround, let's continue the discussion over there (if necessary) so we don't derail this thread.

**Comment by LouisCochinho** [comment]
Hi, I'm facing the same issue with "Definitions/body parameters are not exported since this API references multiple schemas with OpenAPI components as opposed to one." I have an Api in Api management with a few operations in it. An openapi schema is attached to this api and I can access its definitions. 

But when I want to import (via template ARM) a second schema attached to this api in order to add a new definition, the deployment succeeded but I can't see my newly created definition in the list. The open api specification no longer contains the definitions. Instead it contains a property on the end saying

 ` "x-ms-export-notes": ["Definitions/body parameters are not exported since this API references multiple schemas with OpenAPI components as opposed to one. Schemas referenced: xxx, yyy"]`

**Comment by tquady** [comment]
We are also seeing this issue...  will a fix be worked anytime soon?

**Comment by tquady** [comment]
@RupengLiu @miaojiang Can you please provide an ETA for fix?  Looks like this is tagged as an enhancement which is incorrect as its a bug.  It went from doing, upcoming and then to backlog several times.  

this is causing issues with our deployments to QA and Production APIM blades

**Comment by ThomasBleijendaal** [comment]
I've created a powershell script that reads all schemas within one api and merges all models defined in those schemas into a new one. After that it updates all the operations within that api to use the new schema and delete the old ones. This fixes the broken swagger export and the APIM portal is also working again.

https://gist.github.com/ThomasBleijendaal/e8e0257f8df605ce654c9711c7863073

It works similar to what @bremnes has made, but their script only removes redundant schemas which will not work if you have multiple schemas that are still used by various operations (which will happen when using both Extractor and Creator to generate a single API).

I've tried to fix the tooling instead of using this workaround, but it looks like the APIM ARM deployment will always generate two schemas, so it is hard to fix it using only ARM. My workaround is powershell only, and uses the az cli to invoke rest methods, so one can add a deployment script to the generated ARM templates to fix this after deployment.
